beacon_node/beacon_chain/src/beacon_chain.rs

```diff
@@ -1084,6 +1085,76 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
         }
     }
 
+    /// Check that the shuffling at `block_root` is equal to one of the shufflings of `state`.
+    ///
+    /// The `target_epoch` argument determines which shuffling to check compatibility with, it
+    /// should be equal to the current or previous epoch of `state`, or else `false` will be
+    /// returned.
+    ///
+    /// The compatibility check is designed to be fast: we check that the block that
+    /// determined the RANDAO mix for the `target_epoch` matches the ancestor of the block
+    /// identified by `block_root` (at that slot).
+    pub fn shuffling_is_compatible(
+        &self,
+        block_root: &Hash256,
+        target_epoch: Epoch,
+        state: &BeaconState<T::EthSpec>,
+    ) -> bool {
+        let slots_per_epoch = T::EthSpec::slots_per_epoch();
+        let shuffling_lookahead = 1 + self.spec.min_seed_lookahead.as_u64();
+
+        // Shuffling can't have changed if we're in the first few epochs
+        if state.current_epoch() < shuffling_lookahead {
+            return true;
+        }
+
+        // Otherwise the shuffling is determined by the block at the end of the target epoch
+        // minus the shuffling lookahead (usually 2). We call this the "pivot".
+        let pivot_slot =
+            if target_epoch == state.previous_epoch() || target_epoch == state.current_epoch() {
+                (target_epoch - shuffling_lookahead).end_slot(slots_per_epoch)
+            } else {
+                return false;
+            };
+
+        let state_pivot_block_root = match state.get_block_root(pivot_slot) {
+            Ok(root) => *root,
+            Err(e) => {
+                warn!(
+                    &self.log,
+                    "Missing pivot block root for attestation";
+                    "slot" => pivot_slot,
+                    "error" => format!("{:?}", e),
+                );
+                return false;
+            }
+        };
+
+        // Use fork choice's view of the block DAG to quickly evaluate whether the attestation's
+        // pivot block is the same as the current state's pivot block. If it is, then the
+        // attestation's shuffling is the same as the current state's.
+        // To account for skipped slots, find the first block at *or before* the pivot slot.
+        let fork_choice_lock = self.fork_choice.core_proto_array();
+        let pivot_block_root = fork_choice_lock
+            .iter_block_roots(block_root)
```

Comment:
An alternative to this iterator would be adding `pivot_block_root` to each `ProtoNode`. Although I don't think we should actually do this, it would have the following upsides:

1. No iteration
1. Does not rely upon ancestors still existing in fork choice.

For (1), I don't think this is really very important since iterating involves jumping around in an array via index and that's pretty fast.

For (2), it's safe to assume that ancestors are in fork choice whilst the shuffling look-ahead (2 epochs) is equal to or less than the time it takes to hit finality. There is some argument that adding `pivot_block_root` would remove this assumption and make this function more generic.

I'm not convinced that changing what you have here is strictly better, seeing as it requires more memory to store. I'm happy to leave this as is, just wanted to share some thinking.

Reply:
Yeah, I considered how much of a performance impact the iteration would impose and figured it would be quite minimal compared to everything else happening during block production. If that intuition turns out to be wrong we can always tweak it down the road.

I see your point about (2) as well. I think it will break in quite an obvious way if the shuffling lookahead increases beyond what it is currently (all our blocks will be missing attestations), and we could reconsider our approach if and when that happens (it seems unlikely at this stage).

I'll merge once CI passes (I merged master and tweaked the memoization a tiny bit, for something I noticed while passing over the code with fresh eyes).